h_malloc.c

```diff
@@ -1874,6 +1874,94 @@ EXPORT size_t h_malloc_object_size_fast(const void *p) {
     return SIZE_MAX;
 }
 
+inline void *h_memcpy_real(void *dst, const void *src, size_t len) {
+#if CONFIG_BLOCK_OPS_CHECK_SIZE && !defined(HAS_ARM_MTE)
+    char *p_dst = (char *)dst;
+    char const *p_src = (char const *)src;
+
+    while(len--) {
+        *p_dst++ = *p_src++;
+    }
+
+    return dst;
+#else
+    return memcpy(dst, src, len);
+#endif
+}
+
+inline void *h_memmove_real(void *dst, const void *src, size_t len) {
+#if CONFIG_BLOCK_OPS_CHECK_SIZE && !defined(HAS_ARM_MTE)
+    char *p_dst = (char *)dst;
+    char const *p_src = (char const *)src;
+
+    if(p_src < p_dst) {
+        p_dst += len;
+        p_src += len;
+        while(len--) {
+            *--p_dst = *--p_src;
+        }
+    } else {
+        dst = h_memcpy_real(dst, src, len);
+    }
+
+    return dst;
+#else
+    return memmove(dst, src, len);
+#endif
+}
+
+inline void *h_memset_real(void *dst, int value, size_t len) {
+#if CONFIG_BLOCK_OPS_CHECK_SIZE && !defined(HAS_ARM_MTE)
+    char *p_dst = (char *)dst;
+
+    while(len--) {
+        *p_dst++ = value;
+    }
+
+    return dst;
+#else
+    return memset(dst, value, len);
+#endif
+}
+
+#if CONFIG_BLOCK_OPS_CHECK_SIZE && !defined(HAS_ARM_MTE)
+EXPORT void *h_memcpy(void *dst, const void *src, size_t len) {
```

Comment:
Arguments should be `restrict` pointers.

Also the function should check for overlaps:

```c
if (dst < src + len && dst + len > src) {
    fatal_error("memcpy overlap");
}
```